### ISSUE TITLE
Yolov8: Wrong image size?

### ISSUE THREAD

**Issue by lesept777** [ultralytics/ultralytics#8938]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
I train my model using images of size (320, 320):
`results = model.train(data='./data_div3', epochs=50, imgsz=320, cos_lr=True, lr0=0.002, batch=-1, optimizer='AdamW', augment=False, auto_augment=None)`

After that I run an inference on a few test images. They are generated on the fly from wav files as follows:
```
        sp = librosa.feature.melspectrogram(y=son, sr=sr)
        spec = librosa.amplitude_to_db(sp, ref=1.0)
        spec = (spec - min) / (max - min) * 255
        spec = spec.astype('uint8')
        img = Image.fromarray(spec, mode="L")
        img = ImageOps.colorize(img, black="red", mid="green", white="blue")
        img = img.resize((320,320))
...
        img.save(savefile)
```
Immediately after that, I run the inference:
`pred = model(savefile)`

I get some output as follows, which shows an image size of (640, 640):

> image 1/1 /data1/Yolo/wav_1.jpg: **640x640** 3 0.24, 5 0.19, 8 0.17, 2 0.09, 9 0.08, 2.4ms

If I check the size of the image file wav_1.jpg, it indeed displays 320 x 320.

So does the model convert it to 640 x 640, or is just a weird message that I shouldn't bother about? 
Or should I force the image size in the yaml file of the model somewhere?

```
# Ultralytics YOLO ??, AGPL-3.0 license
# YOLOv8-cls image classification model. For Usage examples see https://docs.ultralytics.com/tasks/classify

# Parameters
nc: 10 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n-cls.yaml' will call yolov8-cls.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.30, 0.20, 1024]
  s: [0.33, 0.50, 1024]
  m: [0.67, 0.75, 1024]
  l: [1.00, 1.00, 1024]
  x: [1.00, 1.25, 1024]

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8  # Passer à 210 !
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 6, C2f, [512, True]]

# YOLOv8.0n head
head:
  - [-1, 1, Classify, [nc]] # Classify
```

BTW, what do all these numbers mean, just after the image size? 

> 3 0.24, 5 0.19, 8 0.17, 2 0.09, 9 0.08, 2.4ms

**Comment by glenn-jocher** [comment]
@lesept777 hello! Great question, and thanks for providing the detailed context. 😊

When you run inference using YOLOv8, the model actually adapts your input image to the default inference size defined in the model’s settings or the size you’ve explicitly set during training or inference (if different). This resizing is to maintain a consistent input size for the model, optimizing the detection process. In your case, even though the original images are at (320,320), they're automatically resized to (640,640) for inference, as suggested by the output message.

To run inference explicitly with your desired image size, you might want to specify the `imgsz` during inference like so:
```python
pred = model(savefile, imgsz=(320, 320))
```
This should help in ensuring that your model uses the size you've intended.

As for the numbers like `3 0.24, 5 0.19, etc.` that follow the image size in the output, they represent detected object classes and their respective confidence scores. For instance, `3 0.24` means class 3 was detected with a confidence score of 0.24.

Hope this clears up any confusion! Feel free to reach out if you have more questions.

**Comment by lesept777** [comment]
Great, I'll try this very soon. Thanks for your answer!

**Comment by glenn-jocher** [comment]
@lesept777 You're welcome! 😊 Glad to help. If you have any more questions or need further assistance as you go along, feel free to reach out. Happy detecting!

**Comment by lesept777** [comment]
I tried it, it works fine. Thanks

**Comment by glenn-jocher** [comment]
@lesept777 Fantastic to hear it worked out for you! 😄 If anything else comes up or you have more questions, we're here to help. Happy coding!

**Comment by lesept777** [comment]
A new question about image size:
Is it possible to tain a model using rectangular images?

I assume that
`results = model.train(data='./data_div3', epochs=50, imgsz=320, cos_lr=True, lr0=0.002, batch=-1, optimizer='AdamW', augment=False, auto_augment=None)`
trains the model using 320 x 320 pixels images.

But how can I declare that my images are, for instance, 64 x 320 pixels?

**Comment by glenn-jocher** [comment]
@lesept777 hey there! 😊 Yes, you can indeed train a model using rectangular images. Just specify `imgsz` with your desired dimensions as a tuple, like this:

```python
results = model.train(data='./data_div3', epochs=50, imgsz=(64, 320), cos_lr=True, lr0=0.002, batch=-1, optimizer='AdamW', augment=False, auto_augment=None)
```

This will train your model using images resized to 64x320 pixels. Happy training!

**Comment by lesept777** [comment]
Thanks a lot Glenn, your my Yolo-angel! :-)

**Comment by glenn-jocher** [comment]
@lesept777 you're very welcome! 😇 Glad I could help. If you have any more YOLO-questions, feel free to shoot them my way. Happy coding!

**Comment by lesept777** [comment]
Actually, when I do this:
`results = model.train(data=f'./data_div3_train{int(train*100)}', epochs=100, imgsz=(32,160), cos_lr=True, lr0=0.002, batch=-1, optimizer='AdamW', augment=False, auto_augment=None, translate=0, scale=1, fliplr=0, erasing=0, crop_fraction=1)`
I get this:

> Ultralytics YOLOv8.1.19 🚀 Python-3.8.10 torch-2.0.1+cu117 CUDA:0 (NVIDIA A100-PCIE-40GB, 40536MiB)
> engine/trainer: task=classify, mode=train, model=yolov8n-medium-small.yaml, data=./data_div3_train20, epochs=100, time=None, patience=100, batch=-1, imgsz=(32, 160), save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train6, exist_ok=False, pretrained=True, optimizer=AdamW, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.002, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0, scale=1, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=None, erasing=0, crop_fraction=1, cfg=None, tracker=botsort.yaml, save_dir=runs/classify/train6
> train: /data1/is156025/fa125436/Confluence/Yolo/data_div3_train20/train... found 120000 images in 10 classes ✅
> val: /data1/is156025/fa125436/Confluence/Yolo/data_div3_train20/val... found 480000 images in 10 classes ✅
> test: None...
> 
>   from  n    params  module                                       arguments
>   0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]
>   1                  -1  1      3504  ultralytics.nn.modules.conv.Conv             [16, 24, 3, 2]
>   2                  -1  1      4176  ultralytics.nn.modules.block.C2f             [24, 24, 1, True]
>   3                  -1  1      5232  ultralytics.nn.modules.conv.Conv             [24, 24, 3, 2]
>   4                  -1  2      7104  ultralytics.nn.modules.block.C2f             [24, 24, 2, True]
>   5                  -1  1      8720  ultralytics.nn.modules.conv.Conv             [24, 40, 3, 2]
>   6                  -1  2     19520  ultralytics.nn.modules.block.C2f             [40, 40, 2, True]
>   7                  -1  1     66570  ultralytics.nn.modules.head.Classify         [40, 10]
> YOLOv8n-medium-small summary: 81 layers, 115290 parameters, 115290 gradients, 1.0 GFLOPs
> TensorBoard: Start with 'tensorboard --logdir runs/classify/train6', view at http://localhost:6006/
> AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
> AMP: checks passed ✅
> **WARNING ⚠️ updating to 'imgsz=160'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'**
> 

So I think that the images are converted to 160 x 160, which is not what I expect.
I'm trying to reduce the FLOPs as much as possible and I wanted to see if there is a difference between image size 32x160 and 160x160.
But if the trainer automatically transforms my images to square images, I can't get what I want.
Can you help me on this ?

**Comment by glenn-jocher** [comment]
@lesept777 hi there! 😊 It sounds like you're exploring some really interesting optimizations with your model!

You're correct in observing that during training, your images get resized to square dimensions based on what you set with the `imgsz` parameter, and as indicated by the warning in your log, YOLO expects an integer value for `imgsz` during training and validation which leads to square resizing.

To work with rectangular images directly as you intend (`32x160`), YOLOv8 doesn't natively support non-square image sizes for training due to how the architecture and data pipeline are designed. However, you can still train with non-square aspect ratios by customizing your dataset preprocessing, ensuring your data fits into the model's expected input shape by padding the shorter side to maintain the aspect ratio.

For reducing FLOPs, focusing on smaller models, adjusting model architecture, or even leveraging techniques like quantization and pruning might be more fruitful avenues. Additionally, optimizing your model post-training with TensorRT (for NVIDIA GPUs), which can take `imgsz` as a list `[h, w]` for `predict` and `export`, might further help streamline deployment efficiency.

But, to directly test with non-square inputs and different aspect ratios purely for inferential purposes or architecture exploration, you'd typically do this during inference or export, adjusting the pipeline accordingly based on the target deployment environment constraints.

If you're keen on experimenting with training directly on non-square resolutions for reducing computational complexity, consider fiddling with model architecture or reaching out to the Ultralytics team for potential updates or workarounds they might be considering.

Hope that sheds some light on your query! Keep up the experimentation, and don't hesitate to reach out for further assistance. 😄
